File: algorithms/medium/gas_station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
      vector<int> diff(gas.size(), 0);
      int total{0};
      for (int i = 0; i < gas.size(); i++) {
        diff[i] = gas[i] - cost[i];
        total += diff[i];
      }
      if (total < 0) return -1;
      int ans{0}, g{0};
      for (int i = 0; i < diff.size(); i++) {
        if (g < 0) {
          ans = i;
          g = 0;
        }
        g += diff[i];
      }
      return ans;
    }
};

File: algorithms/medium/gas_station.rs
impl Solution {
    pub fn can_complete_circuit(gas: Vec<i32>, cost: Vec<i32>) -> i32 {
        let mut diff = vec![0; gas.len()];
        let mut total = 0;
        for i in 0..gas.len() {
            diff[i] = gas[i] - cost[i];
            total += diff[i];
        }
        if total < 0 {
            return -1;
        }
        let mut ans = 0;
        let mut g = 0;
        for (i, d) in diff.into_iter().enumerate() {
            if g < 0 {
                ans = i;
                g = d;
            } else {
                g += d;
            }
        }
        ans as _
    }
}
